language/documentation/book/src/functions.md

```diff
@@ -66,6 +73,82 @@ script {
 }
 ```
 
+#### `public(friend)` visibility
+
+The `public(friend)` visibility modifier is a more restricted form of the `public` modifier to give more control about where a function can be used. A `public(friend)` function can be called by
+- other functions defined in the same module, or
+- functions defined in modules which are explicitly specified in the **friend list** (see [Friends](./friends.md) on how to specify the friend list).
+
+Note that since we cannot declare a script to be a friend of a module, the functions defined in scripts can never call a `public(friend)` function.
+
+```move=
+address 0x42 {
+module M {
+    friend 0x42::N;  // friend declaration
+    public(friend) fun foo(): u64 { 0 }
+    fun calls_foo(): u64 { foo() } // valid
+}
+
+module N {
+    fun calls_M_foo(): u64 {
+        0x42::M::foo() // valid
+    }
+}
+
+module other {
+    fun calls_M_foo(): u64 {
+        0x42::M::foo() // ERROR!
+//      ^^^^^^^^^^^^ 'foo' can only be called from a 'friend' of module '0x42::M'
+    }
+}
+}
+
+script {
+    fun calls_M_foo(): u64 {
+        0x42::M::foo() // ERROR!
+//      ^^^^^^^^^^^^ 'foo' can only be called from a 'friend' of module '0x42::M'
+    }
+}
+```
+
+#### `public(script)` visibility
+
+The `public(script)` modifier is designed to allow module functions to be safely and directly invoked much like scripts. A `public(script)` function can only be called from a *script* context, which is either:
+- the function defined in a transaction script, or
+- another `public(script)` function.
+
+Essentialy, this rule implies that once the execution transitions to a non-script context via a call to any non-`public(script)` function, there is no turning back, i.e., there is no way to call a `public(script)` function again.
+
+```move=
+address 0x42 {
+module M {
+    friend 0x42::N;  // friend declaration
```

Comment:
Should we remove the friend list here? It could be confusing to include it in this example.

Reply:
Removing it will lead to a compilation error when reader tries to copy-paste the code, which, I guess, can be confusing in a different way?

Reply:
So, are you saying that the error when compiling `N::calls_M_foo` will be different if `N` is not on the friend list of `M`? I wouldn't have expected that, since `foo` does not have `friend` visibility.

Regardless, how about just removing module N in this example? You've already got an example of the "can only be called from a script context" error, so adding another one from a separate module doesn't really add much. Otherwise, this seems to suggest that there is something special about friend modules calling script functions, which I don't believe is true.

Reply:
Ah, my bad. I confused this example with the one above in the friend visibility section. Yes, it should be removed.